### ISSUE TITLE
Playback Shuttle control

### ISSUE THREAD

**Issue by pmjdebruijn** [jliljebl/flowblade#738]
Flowblade 2.2 (MLT 6.16)

I've recently gotten a Contour Design ShuttleXpress Black jog wheel, it's a hidraw device, which combined with a bit of software ( https://github.com/pmjdebruijn/shuttleevent ), can "fake" keypresses depending on the application window title.

In general it works quite nicely with increasing numbers of Ctrl+Left/Right mapped to how far the dial is rotated.

I'm noticing however, that sometimes things can go a bit unexpected depending on focus in the application.

For example when actively working with a compositor, like Affine Blend, typically the arrow keys will move the image being composited, as opposed the timeline, unless focus is moved.

And while this makes sense for keyboard+mouse workflow, it feels odd when using a jog wheel, ideally I'd like the jog wheel to always map to timeline changes.

Would you be open to adding a second set of hotkey that are analogous to left+right key, but contextually insensitive, for example ',' and '.' , so when these are pressed, it always affects only the timeline never anything else, same for Ctrl+',' or Ctrl+'.' of course :)

**Comment by dvdlvr** [comment]
For info, with the move to Python 3 I have support of jogshuttles on my development list for Flowblade, using the USB interface. This wasn't possible with python 2 because it had no interrupt handling. It will take a while to develop, but I have a jog shuttle myself so I really want to use it again (like I did with kdenlive). I would restrict the jog shuttle use to the timeline and preview window. Sent from my Huawei tablet

**Comment by pmjdebruijn** [comment]
That's mostly a hardware specific endevour though. 

The nice thing about the very affordable ShuttleXpress Black is that through keyboard emulation it can work (to some extent) with anything, with very little modification. Having contextually insensitive timeline hotkeys would make my shuttle work near perfectly, with (hopefully) much less effort required. And these hotkeys may be useful for normal keyboard+mouse users at some point as well.

**Comment by jliljebl** [comment]
I don't have related hardware and I'm quite fine mouse + keys workflow so I can't do really patches for this one.

For accepting patches on this I'd would say that what I look for is that the new patches do not interfere with mouse+keyboard workflow nor create ugly or complicated interface 

E.g. just simply making  ',' and '.' always do playback actions would stop them from working when user is writing Range Log comments, which is only (just) acceptable as opt in. But presenting that opt in for user is actually a non-trivial design issue.

I think it would best if a clear design would be presented with info on what will added where on GUI front and which exact keys do exactly what with playback control. Then it would be possible to agree on exact design before any coding

**Comment by pmjdebruijn** [comment]
I don't quite understand the relevance to the range log (I haven't used that feature yet)? The idea is to have the arrow keys still work exactly like they do now. Just add ',' '.' for timeline only movement (regardless of which display element is active). Or are ',' and '.' already used for something?

**Comment by jliljebl** [comment]
',' '.' Already have 

 if ',' '.' contgext in sensitive

**Comment by pmjdebruijn** [comment]
Keep in mind I'm just using ',' and '.' as examples... Basically any two yet unmapped keys would suffice...

**Comment by jliljebl** [comment]
Yes  ',' and '.' are already used, see "Edit"->"Keyboard Preferences"

Range Log is just example of text input.

If  ',' '.' are not context sensitive you  can either get "............................" or ",,,,,,,,,,,,,,,,,,,,,,," into text input boxes from Jog Wheel or  ',' '.'  doing playback actions when user is writing text into text entries. Tthese kinds of input side effects are not acceptable unless user specifically chooses this behavior.

If we can find keycodes that are never typed that would be better but this feels Strongly like a feature needing to be explicitly activated.

**Comment by dvdlvr** [comment]
The way I see it any HID device, like Pascal's ShuttleXpress and my Shuttle Pro, would be able to work with my code (once I make it) to full extent. No need for os-level key mapping, but native support. It would allow the hard key mapping as present in kdenlive for the Shuttle. It would be more work to implement than ask users to do os key mapping but should be more user-friendly, without the need of additional software. And I'm going to do it anyway, because I want this solution for me 😊Sent from my Huawei tablet

**Comment by pmjdebruijn** [comment]
@dvdlvr, having native support would be nice to have, but it's not a single final solution either. I need the shuttleevent daemon to be able to use the shuttle with applications which do not have native support for it (basically everything), and once that daemon is running, the input event device is locked, so any applications with native support won't be able to open the device anyhow. Which is why I would heavily prefer a hotkey based solution.

@jliljebl, good point, but for example the Fx keys seem mostly unmapped...

**Comment by dvdlvr** [comment]
@pascal - that sounds like a very specific use case, not like something that the general user will be concerned with. Sent from my Huawei tablet

**Comment by jliljebl** [comment]
Well, there is plenty of room in "Preferences"->"Playback" panel for everyone to get their preferred workflow (if the patches come in with reasonable quality), that's why it is there...

@pmjdebruijn F11 is used and I use F12 constantly for to launch different event on development, but we could give couple with **Control**  key, but those would not go into Keyboard Preferences window, but presented as checkbox in  "Preferences"->"Playback".

**Comment by ratherlargerobot** [comment]
After reading this thread, I ordered a Contour ShuttleXpress to play with. I whipped up a little proof of concept Python script that uses pyusb, and have gotten as far as reading and decoding input from the device. That is to say, I can make sense of the jog, shuttle, and five buttons, and it's at the point where I can print stuff out to stdout based on any input on the device.

@dvdlvr, how far have you gotten with this? Maybe we could combine our efforts?

**Comment by ratherlargerobot** [comment]
An update:

I now have a (really hacky) prototype integration working. The jog and shuttle controls the Flowblade timeline! I also hacked the third (middle) button to do play/pause.

At the moment, the prototype is an unholy mess. I wouldn't submit a pull request for it in this form. But I believe it can be polished up, and wanted to talk about how something like this should integrate into Flowblade.

My proof of concept works by having Flowblade depend on pyusb, and then taking direct control of the USB device. From there, it registers a recurring GTK callback function that periodically handles USB input.

Handling input is divided into several parts:
- Decoding the raw bytes from the USB device
- Updating a model of the device (e.g. detecting key press change events, jog wheel direction, etc.)
- Having the model handle callback functions, which ultimately control the timeline

There are so many problems with what I have right now on my computer:
- It just assumes that pyusb is present, rather than making it a soft dependency
- It just assumes that a Contour ShuttleXpress USB device is available at all times
- It should probably be using interrupts as @dvdlvr suggests instead of GTK callbacks
- The shuttle still needs some debouncing and general tuning
- It should be possible to support more than one kind of USB HID interface
- It should be possible to map the buttons to various things according to user preferences

So, I did the stupid, simple, direct thing. But integrating something like this properly into the program would take some more design work, including careful thought about the user interface.

Presumably there should be something in the preferences menu where you have to go and explicitly turn on support for the device. And probably more preferences stuff where you can choose where to map the keys, and select other devices. There should be at least some minimal level of driver architecture, to make it easy to add support for additional devices later.

By the way, it turns out that controlling the Flowblade timeline with the jog/shuttle wheel is more fun than I thought! I bought this device for a USB hacking project as much as anything, but it reminds me of editing on those old professional Sony VTRs...

@jliljebl what are your thoughts on all of this?

**Comment by dvdlvr** [comment]
@Nathan, I haven't worked on this yet on python3. I am currently 
finalising the first phase of subtitle support.
But yes, I agree that this needs to be properly structured and open to 
all sorts of HID devices.

I would suggest using multiple xml files for configuration - possibly 
with an input screen in the preferences dialog somewhere - to keep 
things simple.
As not everybody had a shuttle type device it doesn't seem to be right 
to place preferences in the 'normal' preferences storage. It'll be a bit 
like the shortcut files in the res/shortcuts directory. This also allows 
development to be rather simple as all you need to change for testing is 
the xml file - the preferences possibility is not needed until we hand 
it to the users.
It will make it easy to distribute multiple device specifics between 
systems and users without changing the Flowblade code and without 
risking crashes because of incompatible Preferences.
In the xml we could add the vendor and product id of the device we're 
looking for, thus avoiding hardcoding it.
We can also code the 'keystroke' codes generated by each device, give 
them a proper name and link them to an action. We will have to code the 
available actions in Flowblade, of course.
This way the decoding of the raw bytes of the device is a matter of 
comparing it with the strings we've put in the xml file and mapping it 
to the internally coded action that it links to in the xml file.

The big problem with pyusb, as you will have noticed, is that you need 
an infinite loop of some sort to constantly poll the input. This would 
be a waste of resources, which is why I never went forward with it on 
python2. Python3 supplies better usb interface capability, from what I 
remember when investigating it.
A hook / interrupt method would be more efficient and easier to 
integrate into the existing code.

As a side note, in case you didn't know, you can vary the speed of the 
timeline in various ways, using the 'jkl' method with harcoded speed 
progression and using the arrow keys. The user preferences / playback / 
fast forward... options allow you to set speed accelerators for 
different dead keys, combined with the arrow keys. E.g. set the Ctrl 
value to 3 and the Shift value to 2 and you have 1x, 2x, 3x, 6x 
(Ctrl+Shift) speed with the arrow keys. I submitted that when I saw I 
wouldn't be able to do the Shuttle code until py3.

**Comment by ratherlargerobot** [comment]
@dvdlvr Where was this stuff you found about Python 3 having improved support for handling USB interrupts?

**Comment by ratherlargerobot** [comment]
@dvdlvr The XML files for devices is a good idea. Along with some way to select them that doesn't get in the way of normal use of the application for the majority of users.

I did see the timeline speed adjustments that you added while poking around the code, very cool!

In the spirit of knowledge sharing, here are some assorted facts I learned about this device while experimenting with it:

```
$ lsusb | grep Contour
Bus 001 Device 048: ID 0b33:0020 Contour Design, Inc. ShuttleXpress
```
The OS has some sort of default driver for it, that maps the shuttle operations to the equivalent of the middle scroll wheel on a mouse.

On Ubuntu 18.04 I had to add a file like this in order to be able to take control of the device as a non-root user. This might not be the most optimal setting, but it's what I was using during experimentation that actually worked:

```
# /etc/udev/rules.d/90-shuttleexpress.rules

ATTRS{idVendor}=="0b33", ATTRS{idProduct}=="0020", SYMLINK+="shuttleexpress", MODE="0666"
```

Once you start reading data from the device using pyusb, it returns data as a five-element array of integers. The values it returns can be interpreted as follows:

**data[0]** shuttle position. 0 is in the middle, 1-7 is moving increasingly farther to the right, 255-249 is moving increasingly farther to the left

**data[1]** jog wheel rotary encoder. When the jog wheel is rotated to the right, the numbers increase by one. When the jog wheel is rotated to the left, the numbers decrease by one. However, it is an 8-bit unsigned int, and when you go higher than 255 or lower than 0, it wraps around to the other side. This has to be detected and accounted for.

**data[2]** unused

**data[3]** encoding of buttons 1-4. Bit masking can pull out the buttons from the int. Button 1 is 16, 2 is 32, 3 is 64, and 4 is 128.

**data[4]** encoding of button 5. Button 5 is 1

At least when polling for data, new data is only available when something changes. If you rotate the jog wheel, data is returned. If you press a button, data is returned. If you release a button, data is returned. If you move the shuttle to a new position, data is returned.

The jog wheel and buttons are pretty easy to deal with. Just keep track of the current and previous state of each, and you can detect changes. You also get data when the buttons are released, so you can detect press and release events. All five buttons can be pressed or not pressed in any combination simultaneously, and all of the data comes through.

The shuttle requires a different approach. You only get data when the position of the shuttle changes. And if you're polling for data, you might miss some of the transitions! For example, if you move the shuttle all the way to the right, you'll get one set of data returned for each step along the way. But then once it's all the way to the right, you don't continue to get data each time you poll, you are just supposed to remember what the last value is. That means that if you want it to act like a shuttle, you need to remember the last shuttle setting you saw, and then have some other code that continues to, say, run forward through the timeline. In my prototype code, I just had the GTK polling code also check the most recent saved value that we saw from the shuttle, and keep moving the timeline in that direction until the shuttle went back to zero. But unlike all the other inputs, it has to perform this behavior on a recurring basis until further notice, which is different than the jog wheel and the buttons. I'm not quite sure where to put this sort of recurring behavior in the real program.

Another challenge is that if you have the shuttle held, say, all the way to the right and then let go of it, it will spring back a bit past zero before it centers. If you combine this will polling and maybe missing the last state change from the device, this means that the last thing you see from the device could be an instruction to go slightly in reverse, with no indication that it ever reached zero. Even polling every 5 milliseconds didn't fix this at all.

Currently, a lot of the transport control code is tied in with handling key events. A typical block of code that controls timeline behavior will have a bunch of if checks to see if certain windows are selected and certain keys are pressed. And then only after all of those conditions are met, there will sometimes be a series of calls that will take into account whether the program is in trim mode, etc. before finally controlling the timeline in a context-sensitive way.

If jog/shuttle USB HID interfaces are to be supported, it seems like a lot of the logic about how to move the timeline in different timeline modes should be moved out of keyevents.py into a central location. Then keyevents.py and the USB HID code could both call abstract functions or methods to control the timeline, free of any assumptions about mouse focus or USB device status. In my prototype code I simply wrote my own little duplicate functions with names like play_pause(), move_timeline(frames), etc. and mapped the USB handling code to call these as callbacks when the state changed. And then inside of my functions, I basically copy and pasted some of the bits from keyevents.py (but stopped short of adding all the code to move the jog wheel in trim mode and such).

I would really like to learn more about Python 3 USB interrupt handling, and how you see that working. Despite doing a fair amount of searching, I didn't really come across anything that looked promising.

**Comment by ratherlargerobot** [comment]
Were you looking at the python-libusb1 module?

**Comment by dvdlvr** [comment]
I really can't remember. I last looked at it 3 years ago. I couldn't find any notes. I will have to dig deeper to try and find anything. Sent from my Huawei tablet

**Comment by ratherlargerobot** [comment]
I played around with python-libusb1, and got a little toy script going that decodes data using that library. Despite having method names with words like "interrupt", everything still seems to require polling at some level. On the plus side, the libusb1 toy script seemed a lot better about actually receiving the last value from the shuttle, although it still had to poll for updates.

I also looked around the Flowblade code some more, to see how the JKL navigation works. It turns out that the JKL timeline navigation works by calling an mlt Producer object, and telling it to play at a certain speed. That's a neat trick, because then the MLT Producer can just go and handle playback in its own thread. I then tried to add a playback speed that was less than 1.0, which didn't seem to work. The hope was to just have the MLT Producer handle slower than 1x playback for the lower settings on the shuttle, but it just didn't work. Even if you tell the MLT Producer to use a playback speed of 0.9, it just never advances to the next frame.

**Comment by ratherlargerobot** [comment]
I think I could probably make this work reasonably well with the polling model. However, I don't want to go and do all that work if there's a better way to do this, or if it might not get merged into the program. I'm going to take a break from this for a while.

**Comment by dvdlvr** [comment]
Thanks for looking into this so far. I'll finish my phase 1 subtitle support and then take another look. Sent from my Huawei tablet

**Comment by pmjdebruijn** [comment]
> @pascal - that sounds like a very specific use case, not like something that the general user will be concerned with.

I don't follow? What use-case?

The shuttlevent repo basically allows the shuttle to be setup to work with any random application dynamically, given that it monitors which window is active, so different applications get have different keymaps, which makes such a shuttleevent daemon quite preferable to native application support, as enabling native support means the daemon would need to be shutdown for that to work.

On a sidenote, I'm working on my shuttleevent repo again:
https://github.com/pmjdebruijn/shuttleevent

**Comment by jliljebl** [comment]
Closing this as not active. It is a useful feature and we may add it someday, but right now things do not seem to be moving.

**Comment by ratherlargerobot** [comment]
I have added jog/shuttle support to Flowblade! Better late than never.

https://github.com/jliljebl/flowblade/pull/1052
